README.md

```diff
@@ -2,15 +2,112 @@
+A new message is always added to at the end of the template class. Its id is one more of the message above it.  
+
+- retryable
+
+If message will do a read-only or idempotent task on the server we mark it as retryable = true, otherwise false.
+
+- response
+
+Type of the response of this message from server.  An existing type can be chosen from / or a new one can be added.
+Following two classes contains response types and definitions. 
+
+    /hazelcast/src/main/java/com/hazelcast/client/impl/protocol/constants/ResponseMessageConst.java
+    /hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ResponseTemplate.java
+
+Note that return value of `exampleAddListener` is Object. This type is not used when generating protocol. 
+
+
+- event
+
+This field is optional. And it is list of event types. 
+If message is a listener registration, and sends events, event types is described here. 
+
+An existing type can be chosen from / or a new one can be added.
+Following two classes contains event types and definitions. 
+
+    /hazelcast/src/main/java/com/hazelcast/client/impl/protocol/constants/EventMessageConst.java
+    /hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/EventResponseTemplate.java
+
+- Since tag on message 
+
+A message itself or its parameters can have since tag. If tag is omitted from message it means @Since("1.0")
+
+Version should be given accordingly to the version of protocol. Protocol version can be found in master pom.xml
+
+Example : <version>1.6.0-13</version>
+
+This means that next release of protocol will be 1.6. Since tag should be as follows `@Since("1.6")`
+
+- Since tag on parameters
+
+New parameters can be added to a message. It is always added at the end of parameter list.
+
+Old parameters can not be deleted or their types can not be modified. 
+
+When adding a new parameter, if version we add new parameter is same as message introdued @Since tag can be omitted.
+If parameter is added in a new version, then @Since tag should be added in front of parameter type. 
+
+- name of the message
+
+Name of the message codec is given via function name . It is `exampleAddListener` in this example.
+
 ## Output directory of generated classes
 
     hazelcast-client-protocol/hazelcast/target/generated-sources/annotations
 
-## To generate java classes and hazelcast-client-protocol.jar
+## To generate java classes and hazelcast-client-protocol.jar 
 
     mvn clean compile
 
-We have code generation for some other languages as well. 
+Code generation needs to download an existing hazelcast and compile against it. Following properties from master
+`pom.xml` are used to find a hazelcast branch. 
+
+    <hazelcast.git.repo>hazelcast</hazelcast.git.repo>
+    <hazelcast.git.branch>master</hazelcast.git.branch>
+
+If you want to generate code against a specific branch of yours, replace them accordingly as follows:
```

Comment:
This is especially useful when you modify (add field to) an existing message. In that case, this repo can not be compiled against hazelcast/master, because generated codecs has a field missing in some methods.  In that case, you need to put your development branch here, so that we can verify generated codecs and development branches are consistent.